discovery/consul/consul.go

```diff
@@ -414,103 +420,100 @@ func (d *Discovery) watchService(ctx context.Context, ch chan<- []*targetgroup.G
 	}
 
 	go func() {
+		ticker := time.NewTicker(d.refreshInterval)
 		var lastIndex uint64
-		for range ticker.C {
-			srv.watch(ctx, ch, &lastIndex)
+		catalog := srv.client.Catalog()
+		for ; true; <-ticker.C {
+			srv.watch(ctx, ch, catalog, &lastIndex)
 		}
-	}
+		<-ctx.Done()
+		ticker.Stop()
+	}()
 }
 
-// Continuously watch one service.
-func (srv *consulService) watch(ctx context.Context, ch chan<- []*targetgroup.Group, lastIndex *uint64) {
-	catalog := srv.client.Catalog()
-
-	lastIndex := uint64(0)
-	for {
-		level.Debug(srv.logger).Log("msg", "Watching service", "service", srv.name, "tag", srv.tag)
-
-		t0 := time.Now()
-		nodes, meta, err := catalog.Service(srv.name, srv.tag, &consul.QueryOptions{
-			WaitIndex:  lastIndex,
-			WaitTime:   watchTimeout,
-			AllowStale: srv.discovery.allowStale,
-			NodeMeta:   srv.discovery.watchedNodeMeta,
-		})
-		elapsed := time.Since(t0)
-		rpcDuration.WithLabelValues("catalog", "service").Observe(elapsed.Seconds())
+// Get updates for a service.
+func (srv *consulService) watch(ctx context.Context, ch chan<- []*targetgroup.Group, catalog *consul.Catalog, lastIndex *uint64) error {
+	level.Debug(srv.logger).Log("msg", "Watching service", "service", srv.name, "tag", srv.tag)
 
-		// Check the context before potentially falling in a continue-loop.
-		select {
-		case <-ctx.Done():
-			return
-		default:
-			// Continue.
-		}
-
-		if err != nil {
-			level.Error(srv.logger).Log("msg", "Error refreshing service", "service", srv.name, "tag", srv.tag, "err", err)
-			rpcFailuresCount.Inc()
-			time.Sleep(retryInterval)
-			continue
-		}
-		// If the index equals the previous one, the watch timed out with no update.
-		if meta.LastIndex == lastIndex {
-			continue
-		}
-		lastIndex = meta.LastIndex
-
-		tgroup := targetgroup.Group{
-			Source:  srv.name,
-			Labels:  srv.labels,
-			Targets: make([]model.LabelSet, 0, len(nodes)),
-		}
-
-		for _, node := range nodes {
+	t0 := time.Now()
+	nodes, meta, err := catalog.Service(srv.name, srv.tag, &consul.QueryOptions{
+		WaitIndex:  *lastIndex,
+		WaitTime:   watchTimeout,
+		AllowStale: srv.discovery.allowStale,
+		NodeMeta:   srv.discovery.watchedNodeMeta,
+	})
+	elapsed := time.Since(t0)
+	rpcDuration.WithLabelValues("catalog", "service").Observe(elapsed.Seconds())
+
+	// Check the context before in order to exit early.
```

Comment:
It's more usual to do this up in the for

Reply:
I think this was originally meant to exit before handling the response or the error.
I'm happy to remove it though.